expand README with explanation of what it does and how to use it

davewhiteland opened the pull request:
Closes #1 

It would be good if the README included at least the following:

- [x] Explanation of what this script does
- [x] How to use the latest version (by having `master` in the URL)
- [x] How to use a specific version (e.g. with SHA or tag in the URL)

Commit:
add explanation of what this script does

Commit:
add details for targeting specific versions in the curl URL

tmtmtmtm commented on README.md:
```diff
@@ -1,9 +1,71 @@
+you create a pull request: once on the branch you’ve been working on, and once
+on the branch you’re targeting (commonly, this will be `master`). The tests on
+your branch *must pass* (of course)... but tests on the target branch *must
+fail*.
```
I don't think this gets across well enough that it's the *tests from your branch* that will be running on the target. Saying "tests on the target branch" is on the wrong side of ambiguous there, I fear, especially as this approach isn't really a common concept.

tmtmtmtm commented on README.md:
```diff
@@ -1,9 +1,71 @@
+This will run v0.1.0 of the script (see [all versions available](https://github.com/everypolitician/ensure-regression-tests/releases)).
+
+If you want to use a specific version of the script, or perhaps always want the
+latest one, be explicit in the URL you supply to `curl`. That is, you can use
```
I think I'd switch the order of these around. We want to promote simply using the latest release by default, but let people know how to pin to a specific version if they really want.

tmtmtmtm commented on README.md:
```diff
@@ -1,9 +1,71 @@
+If you’re adding a new feature, you’d expect to add new tests for that new
+feature. You’d also expect those to fail because by definition the feature your
+tests are testing isn’t on the target branch (yet) — if it is, you’re not
+adding a new feature, you’re refactoring a new one (see below).
```
"refactoring a new one" is slightly odd terminology here. I suspect this was just a thinko on "new".

tmtmtmtm commented on README.md:
```diff
@@ -1,9 +1,71 @@
+If you’re fixing a bug, that bug snuck in despite the tests all passing on the
+target branch. So you’ve certainly got to add new tests to catch this bug on
+your (pull request’s) branch, and those tests must fail on the target branch
+(because if they didn’t, it wouldn’t be a bug, would it?). So
```
This scenario is the key to the whole approach, so I think it's worth being a bit more explicit here about the underlying problem (and probably making it the first in the list). "Despite the tests all passing on the target branch" is certainly correct, but I fear it's just a little bit too subtle, and I'm finding the "(because if they didn't, it wouldn't be a bug)" quite difficult to follow. The basic principle here is: 

1. You've found a bug
2. Your existing tests didn't catch that bug
3. So when fixing that bug, you add a test
4. *BUT* … that test has to fail against the previous code, or else it's not actually testing what you think.

tmtmtmtm commented on README.md:
```diff
@@ -1,9 +1,71 @@
+
+Here’s the override case. If you’re just refactoring — you’re not adding a
+feature, you’re not fixing a bug — then you don’t want to add new tests: you
+emphatically want to show that you haven’t broken the existing ones.
```
The "emphatically" here is puzzling to me. This implies that this script is going to do something here to give you more surety. But all we're doing is bailing out.

Commit:
make README clearer which tests it's running (pull request branch)

Addresses [https://github.com/everypolitician/ensure-regression-tests/pull/2#discussion_r105137414](@tmtmtmtm's comment).

Commit:
put usage example that targets `master` first

Addresses:[https://github.com/everypolitician/ensure-regression-tests/pull/2#discussion_r105137648](@tmtmtmtm's comment).

Commit:
correct typo/thinko: "refactor[...] existing one" not a new one

Addresses [https://github.com/everypolitician/ensure-regression-tests/pull/2#discussion_r105137867](@tmtmtmtm's comment).

Commit:
rewrite bug-fix scenario to make it the emphasis, and show it first

Addresses [https://github.com/everypolitician/ensure-regression-tests/pull/2#discussion_r105139465](@tmtmtmtm's comment).

Commit:
rewrite desription of refactor branch to clarify script does nothing

Addresses [https://github.com/everypolitician/ensure-regression-tests/pull/2#discussion_r105139671](@tmtmtmtm's comment).
